sdk/mobile-center-updates/src/main/java/com/microsoft/azure/mobile/updates/Updates.java

```diff
@@ -0,0 +1,971 @@
+import static com.microsoft.azure.mobile.updates.UpdateConstants.PARAMETER_PLATFORM;
+import static com.microsoft.azure.mobile.updates.UpdateConstants.PARAMETER_PLATFORM_VALUE;
+import static com.microsoft.azure.mobile.updates.UpdateConstants.PARAMETER_REDIRECT_ID;
+import static com.microsoft.azure.mobile.updates.UpdateConstants.PARAMETER_RELEASE_HASH;
+import static com.microsoft.azure.mobile.updates.UpdateConstants.PARAMETER_REQUEST_ID;
+import static com.microsoft.azure.mobile.updates.UpdateConstants.PREFERENCE_KEY_DOWNLOAD_ID;
+import static com.microsoft.azure.mobile.updates.UpdateConstants.PREFERENCE_KEY_DOWNLOAD_URI;
+import static com.microsoft.azure.mobile.updates.UpdateConstants.PREFERENCE_KEY_REQUEST_ID;
+import static com.microsoft.azure.mobile.updates.UpdateConstants.PREFERENCE_KEY_UPDATE_TOKEN;
+import static com.microsoft.azure.mobile.updates.UpdateConstants.SERVICE_NAME;
+
+/**
+ * Updates service.
+ */
+public class Updates extends AbstractMobileCenterService {
+
+    /**
+     * Shared instance.
+     */
+    @SuppressLint("StaticFieldLeak")
+    private static Updates sInstance = null;
+
+    /**
+     * Current login base URL.
+     */
+    private String mLoginUrl = DEFAULT_LOGIN_URL;
+
+    /**
+     * Current API base URL.
+     */
+    private String mApiUrl = DEFAULT_API_URL;
+
+    /**
+     * Application context, if not null it means onStart was called.
+     */
+    private Context mContext;
+
+    /**
+     * Application secret.
+     */
+    private String mAppSecret;
+
+    /**
+     * If not null we are in foreground inside this activity.
+     */
+    private Activity mForegroundActivity;
+
+    /**
+     * Remember if we already opened browser to login.
+     */
+    private boolean mBrowserOpened;
+
+    /**
+     * In memory token if we receive deep link intent before onStart.
+     */
+    private String mBeforeStartUpdateToken;
+
+    /**
+     * In memory request identifier if we receive deep link intent before onStart.
+     */
+    private String mBeforeStartRequestId;
+
+    /**
+     * Current API call identifier to check latest release from server, used for state check.
+     * We can't use the ServiceCall object for that purpose because of a chicken and egg problem.
+     */
+    private Object mCheckReleaseCallId;
+
+    /**
+     * Current API call to check latest release from server.
+     */
+    private ServiceCall mCheckReleaseApiCall;
+
+    /**
+     * Latest release details waiting to be shown to user.
+     */
+    private ReleaseDetails mReleaseDetails;
+
+    /**
+     * Last update dialog that was shown.
+     */
+    private AlertDialog mUpdateDialog;
+
+    /**
+     * Current task inspecting the latest release details that we fetched from server.
+     */
+    private DownloadTask mDownloadTask;
+
+    /**
+     * Current task to process download completion.
+     */
+    private ProcessDownloadCompletionTask mProcessDownloadCompletionTask;
```

Comment:
This is a task that does something *after* downloaded. We can have a better name but it doesn't come up with a better name. Let's keep it for now.